Created `has_autograded_submissions` to detect whether student group has at least one non-rejected AG submission

Normanwqn opened the pull request:
To facilitate fixing this [bug](https://github.com/eecs-autograder/autograder.io/issues/31) where we needed an efficient way to tell whether a student group's has submitted any non-rejected submissions, we introduced a new field `has_autograded_submissions` in data returned by `ListHandgradingResultsView` Django Rest API. `has_autograded_submissions` is true if the student group has at least one submission on the autograder whose status is `ag_models.Submission.GradingStatus.finished_grading`, and false otherwise, when the student group has no submissions or all their submissions have been rejected.

Previous unit tests have been modified to accommodate this change. New unit tests and interfaces have been written to test student groups who only have rejected submissions.

Commit:
Provide has_autograded_submissions for each group with handgrading result list

Addresses https://github.com/eecs-autograder/ag-website-vue/pull/509#issuecomment-2066732797

Commit:
Add missing Submission import

Normanwqn committed:
Use any generator in finished_grading_result_exist

Normanwqn committed:
Fix test to accomodate has_autograded_submissions
Fix extraneous Dockerfile
Fix existing test cases to pass with additional API  'has_autograded_submissions'

Normanwqn committed:
Delete import of grading status

Normanwqn committed:
Add interface & tests for only rejected groups

Compatible with previous test cases

Where
[0, num_results) -> groups with accepted submissions and handgraded
results

[num_result, num_results + num_groups_only_reject) -> groups with only
rejected submissions and no handgraded results

[num_result + num_groups_only_reject, num_groups): groups with no
submissions and no handgraded results

james-perretta commented on autograder/handgrading/views/handgrading_result_views.py:
```diff
@@ -332,6 +332,15 @@ class ListHandgradingResultsView(AGModelAPIView):
                                             },
                                         }
+                                    },
+                                    'has_autograded_submissions': {
```
This is just a bikeshedding question, but perhaps we could call this something like "has_handgradeable_submission" instead?

Normanwqn replied:
Fixed

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -698,14 +698,34 @@ def test_non_staff_non_handgrader_get_handgrading_results_permission_denied(self
     def do_handgrading_results_test(self, user: User, *,
-                                    num_results: int, num_groups: int = None,
+                                    num_results: int,
+                                    num_groups_only_reject: int = 0,
```
I'd like to suggest a slightly different way of going about this: Could we change this parameter to `submission_statuses: list[ag_models.Submission.GradingStatus] | None = None`? When none, populate it as a list of "finished grading", otherwise assert its length equals num_groups. Then when creating the submissions, use the statuses from the list.

The benefit of this is that then we can test it with different non-finished-grading statuses. The important ones besides rejected are `removed_from_queue` and `error`. Let me know what you think, happy to discuss.

Normanwqn replied:
So is the idea to write this method and hence other testing methods to accommodate this change?

There are currently these grading statuses
```
 class GradingStatus(models.TextChoices):
        # The submission has been accepted and saved to the database
        received = 'received'

        # The submission has been queued is waiting to be graded
        queued = 'queued'

        being_graded = 'being_graded'

        # Non-deferred test cases have finished and the group can submit
        # again.
        waiting_for_deferred = 'waiting_for_deferred'
        # All test cases have finished grading.
        finished_grading = 'finished_grading'

        # A student removed their submission from the queue before it
        # started being graded.
        removed_from_queue = 'removed_from_queue'

        # The setup command failed for an AGTestSuite marked as
        # "reject_submission_if_setup_fails".
        rejected = 'rejected'

        # Something unexpected occurred during the grading process.
        error = 'error'
```
Is the idea to accommodate for future testing?

james-perretta replied:
What do you mean by other testing methods? If possible, I'd like this PR to include tests for the `rejected`, `removed_from_queue` and `error` statuses. Let me know if it would be helpful to see some kind of example of what that would look like.

Normanwqn replied:
I am wondering how we are going to test groups who do has not submitted to the AG, shall we use `list[ag_models.Submission.GradingStatus | None] | None` as the type for `submission_statuses`?

By testing methods, I meant if we are changing `do_handgrading_results_test`, we may need to rewrite how other test methods in `ListHandgradingResultsViewTestCase` call `do_handgrading_results_test`. This may take some account of work. 

For `removed_from_queue` and `error` submission, should they have the same behaviour as the `rejected` one? Do we need to test for `received`, `queued`, or `being_graded`?

Normanwqn committed:
Change 'has_autograded_submissions' to 'has_handgradeable_submission'

Normanwqn committed:
Use submission_statuses for future extensions

As per comment
https://github.com/eecs-autograder/autograder-server/pull/680#discussion_r1581527284,
instead of using numbers, we used different submission statuses to
indicated to streamline the interface for future extensibility

Fixed style for modified changes

Normanwqn commented:
@james-perretta I implemented a draft change to reflect what was mentioned [here](https://github.com/eecs-autograder/autograder-server/pull/680#discussion_r1581527284)

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -670,47 +670,65 @@ def test_get_handgrading_results_exclude_staff(self):
+
+    def do_handgrading_results_test(self, user: User, *, num_groups: int = None,
+                                    submission_statuses:
+                                    list[ag_models.Submission.GradingStatus | None] | None = None,
```
I like the direction this is going. Rather than an overloaded use of None (where it can tell us something about the submission or the handgrading result), maybe we should come up with a class to encode this? Maybe something like:

```python
@dataclass
class _GroupHGDataConfig:
    submission_status: ag_models.Submission.GradingStatus = ag_models.Submission.GradingStatus.finished_grading
    has_handgrading_result: bool = True

    def __post_init__(self):
        if has_handgrading_result:
            assert self.submission_status == ag_models.Submission.GradingStatus.finished_grading, 'Submissions with a handgrading result must have finished_grading status'

        if self.submission_status != ag_models.Submission.GradingStatus.finished_grading:
            assert not self.has_handgrading_result, 'Submissions with a status other than "finished_grading" must not have a handgrading result'
```

Then, I think the cases we have to consider are slightly different from the current change. We need to be able to handle:
- A finished_grading submission with a handgrading result
- A finished_grading submission without a handgrading result (has_handgradeable_submission would still be *true* in this case)
- A submission with any status other than finished_grading (and consequently no handgrading result either). has_handgradeable_submission would be false in this case. Note that we don't need to special case rejected for this: you can use `obj_buid.make_submission(group=group, status=status)` to create a submission with the proper status.

After that, we can add copies of the `test_handgrader/staff_get_handgrading_results_only_rejected_submissions` tests for `error` and `removed_from_queue` statuses.

P.S. Sorry I didn't see your question about the type of `submission_statuses` sooner. For some reason I didn't get an email notification about it. In the future, if I don't get back to you within 2 business days, feel free to email me.

Normanwqn replied:
Thanks for your comment, @james-perretta 

I presume we are using `_GroupHGDataConfig` to indicate what kind of submission statuses we are testing and the corresponding `has_handgrading_result`

I am just not sure how we may differentiate these two cases
> * A finished_grading submission with a handgrading result
> * A finished_grading submission without a handgrading result (has_handgradeable_submission would still be true in this case)

by using `_GroupHGDataConfig` alone. 

I presume in the second case, a second group would `make_finished_submission` but the expected `data['handgrading_result']` would be `None` while `data['has_handgradeable_submission']` would still be true. I don't think we currently have this test case, and I add that once I devise some ways to differentiate between `finished_grading` with or without `handgrading_result`.

Normanwqn committed:
Use _GroupHGDataConfig to avoid overloading None

Taking advice from
https://github.com/eecs-autograder/autograder-server/pull/680#discussion_r1591532870,
refactored testing code to more clearly indicate and mark submissions
with no hangrading result with _GroupHGDataConfig(has_handgrading_result
= False, ...) ...

Refactored expected data generation code for clarity

Normanwqn commented:
@james-perretta I implemented another draft to use `_GroupHGDataConfig` to handle test student groups with `finish_grading` status with no handgrading results :)

james-perretta commented:
@Normanwqn Thanks! Taking a look now

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -670,71 +691,107 @@ def test_get_handgrading_results_exclude_staff(self):
+            assert len(submission_configs) == num_groups
+
+        # ensure that finished grading submissions are sorted before rejected submissions
+        submission_configs.sort(
```
Do we need this sorting? As far as I can tell, it shouldn't matter which order the groups are in.
Maybe `submission_configs` wasn't the best suggestion for name--it should probably be something like `group_handgrading_data_configs` instead since we're only making one submission per group.

Normanwqn replied:
Ahh. This was necessary from the old way of generating expected data. I will remove this!

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -670,71 +691,107 @@ def test_get_handgrading_results_exclude_staff(self):
+                data['handgrading_result'] = None
 
+            data['member_names'].sort()
+            data['has_handgradeable_submission'] =\
```
In this case, prefer parentheses for wrapping the line instead of `\`

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -670,71 +691,107 @@ def test_get_handgrading_results_exclude_staff(self):
+                    'total_points_possible': hg_result.total_points_possible,
+                    'finished_grading': hg_result.finished_grading
+                }
+            else:
```
Ah, so looking at what you have, I think this does cover all the conditions I listed, but the way different parts of of that logic are grouped together (some in the if else, some in the assignment to `data['has_handgradeable_submission']`) makes me have to think harder to make sure that mapping is there. Maybe that's a sign that we shouldn't be generating the expected data, but at present I choose to ignore that sign. ;)

To make the mapping between the code and the cases clearer, I'd request you add this condition:
```
elif not config.has_handgrading_result and config.submission_status == ag_models.Submission.GradingStatus.finished_grading:
    data = group.to_dict()
    data['handgrading_result'] = None
    data['has_handgradeable_submission'] = True
```
Similarly, let's explicitly assign has_handgradeable_submission in each branch.

I believe we do already have a test case for this: `test_get_paginated_results_some_groups_have_no_result`

james-perretta commented on autograder/utils/testing/model_obj_builders.py:
```diff
@@ -225,6 +225,14 @@ def make_finished_submission(group: Optional[ag_models.Group] = None,
                            **submission_kwargs)
 
 
+def make_rejected_submission(group: Optional[ag_models.Group] = None,
```
Is this still being used, or can we remove it?

Normanwqn replied:
We can delete this.

Normanwqn committed:
Use clearer case to expected data mapping

Refactored functionality into _GroupHGDataConfig
to have clearer mapping from cases to expected data for each
configuration

Normanwqn commented:
@james-perretta I took some advice and experimented with `match` to hopefully have a clearer mapping between the code and the cases. :)

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -615,6 +616,91 @@ def test_student_delete_result_permission_denied(self) -> None:
+                               rubric: hg_models.HandgradingRubric) -> dict:
+        """Generate the expected data base on submission status and handgrading result status."""
+        submission = self.create_submission(group)
+        match self.has_handgrading_result, self.submission_status:
```
I agree that using match makes sense here. To make it more explicit, I would match on the whole object, which then lets you use keyword matching in the cases. Something like:

```python3
match self:
    case _GroupHGDataConfig(has_handgrading_result=True, self.submission_status=ag_models.Submission.GradingStatus.finished_grading):
       ...
    case _GroupHGDataConfig(has_handgrading_result=False, self.submission_status=ag_models.Submission.GradingStatus.finished_grading):
        ...
    case _GroupHGDataConfig(has_handgrading_result=False):
        ...
    case _:
        assert False, 'Unexpected data configuration'
```

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -615,6 +616,91 @@ def test_student_delete_result_permission_denied(self) -> None:
+        match self.submission_status:
+            # Is handgradable when the submission has a finished grading status
+            case ag_models.Submission.GradingStatus.finished_grading:
+                return self.format_group_data(
```
I appreciate the effort to reduce duplication here, but I do think it's better to have the data generation all in one place in the `generate_expected_data` function instead of having it fragmented. What that would mean here is folding the `hg_result` and `handgrading_data` into the first case in the `match` in `generate_expected_data`.

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -615,6 +616,91 @@ def test_student_delete_result_permission_denied(self) -> None:
+                return self.format_group_data(
+                    group, handgrading_data, has_handgradeable_submission=True)
+            case _:
+                return self.format_group_data(
```
I think this case should be impossible (after applying the other changes I mentioned, I think this particular `match` statement can be removed).

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -615,6 +616,91 @@ def test_student_delete_result_permission_denied(self) -> None:
+    def format_group_data(group, handgrading_result: dict | None,
+                          has_handgradeable_submission: bool) -> dict:
+        """Format the group data for comparison with the response data."""
+        data = group.to_dict()
```
Going along with the comment on `generate_expected_data_with_handgrading_result`, this should be folded into `generate_expected_data`.

james-perretta commented on autograder/handgrading/tests/test_views/test_handgrading_result_views.py:
```diff
@@ -615,6 +616,91 @@ def test_student_delete_result_permission_denied(self) -> None:
 
+@dataclass
+class _GroupHGDataConfig:
+    submission_status: ag_models.Submission.GradingStatus =\
```
Wrap with parentheses instead of `\`

Normanwqn committed:
Merged helper for better readbility&clear mapping

Normanwqn commented:
@james-perretta I merged in the helper functions. You may take a look when you got time :)

james-perretta commented:
@Normanwqn Looks good! Let me figure out why CI isn't running...
In the meantime, make sure that pycodestyle (`pycodestyle autograder`) and mypy (`bash run_mypy.sh`) come up clean if you haven't already.

james-perretta commented:
Going to try closing and reopening to get CI to run.

james-perretta commented:
Looks like that didn't work. @Normanwqn, can you open a fresh PR for this?
